cmd/oras/discover.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras/cmd/oras/internal/option"
+
+	"github.com/need-being/go-tree"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"github.com/spf13/cobra"
+)
+
+type discoverOptions struct {
+	option.Common
+	option.Remote
+
+	targetRef    string
+	artifactType string
+	outputType   string
+}
+
+func discoverCmd() *cobra.Command {
+	var opts discoverOptions
+	cmd := &cobra.Command{
+		Hidden: true,
+		Use:    "discover [options] <name:tag|name@digest>",
+		Short:  "Discover referrers of a manifest in the remote registry",
+		Long: `Discover referrers of a manifest in the remote registry
+
+Example - Discover direct referrers of manifest 'hello:latest' in registry 'localhost:5000':
+  oras discover localhost:5000/hello
+
+Example - Discover all the referrers of manifest 'hello:latest' in registry 'localhost:5000' in a tree view:
+  oras discover localhost:5000/hello -o tree
+
+Example - Discover referrers with type 'test-artifact' of manifest 'hello:latest' in registry 'localhost:5000':
+  oras discover --artifact test-artifact localhost:5000/hello
+`,
+		Args: cobra.ExactArgs(1),
+		PreRun: func(cmd *cobra.Command, args []string) {
+			fmt.Println("Command discover is in preview and might have breaking changes coming.")
```

Comment:
Let's review the UX offline tomorrow.

Reply:
Added `latest` tag a default:

- Tree
```
Using default tag: latest
localhost:5000/net-monitor:latest
├── signature/example
│   └── sha256:2437d8720c28d2bd2ad830d218ebb47942c099e21ff9cf9abbeb5e396203f6dd
│       └── signature/example
│           └── sha256:7c8fa4053dd322a3c1ea069a43c2cfba90ef4db4c9fed71f10da9d3a2071479c
└── sbom/example
    └── sha256:4b6726d63d4283fa579ddf29c4b0838025e2e27e5856e25db78fb50387461ee3
        └── signature/example
            └── sha256:673048d3beb0fada31b268b0e56a89c49fd6c0d2ec9b5ba63330e712b0c6758f
```
- Table
```
Using default tag: latest
Discovered 2 artifacts referencing localhost:5000/net-monitor:latest
Digest: sha256:4d5dd672805a7e20d19d1ae89bf4c5d371c47a018cb33bbc4dec989444602bd0

Artifact Type       Digest
signature/example   sha256:2437d8720c28d2bd2ad830d218ebb47942c099e21ff9cf9abbeb5e396203f6dd
sbom/example        sha256:4b6726d63d4283fa579ddf29c4b0838025e2e27e5856e25db78fb50387461ee3
```
- JSON
```
Using default tag: latest
{
  "referrers": [
    {
      "digest": "sha256:2437d8720c28d2bd2ad830d218ebb47942c099e21ff9cf9abbeb5e396203f6dd",
      "mediaType": "application/vnd.cncf.oras.artifact.manifest.v1+json",
      "artifactType": "signature/example",
      "size": 411
    },
    {
      "digest": "sha256:4b6726d63d4283fa579ddf29c4b0838025e2e27e5856e25db78fb50387461ee3",
      "mediaType": "application/vnd.cncf.oras.artifact.manifest.v1+json",
      "artifactType": "sbom/example",
      "size": 401
    }
  ]
}
```